test/functional/wallet_migration.py

```diff
@@ -0,0 +1,405 @@
+
+        self.log.info("Test migration of a wallet with balance received on the seed")
+        basic2 = self.create_legacy_wallet("basic2")
+        basic2_seed = get_generate_key()
+        basic2.sethdseed(True, basic2_seed.privkey)
+        assert_equal(basic2.getbalance(), 0)
+
+        # Receive coins on different output types for the same seed
+        basic2_balance = 0
+        for addr in [basic2_seed.p2pkh_addr, basic2_seed.p2wpkh_addr, basic2_seed.p2sh_p2wpkh_addr]:
+            send_value = random.randint(1, 4)
+            default.sendtoaddress(addr, send_value)
+            basic2_balance += send_value
+            self.generate(self.nodes[0], 1)
+            assert_equal(basic2.getbalance(), basic2_balance)
+        basic2_txs = basic2.listtransactions()
+
+        # Now migrate and test that we still see have the same balance/transactions
+        basic2.migratewallet()
+        assert_equal(basic2.getwalletinfo()["descriptors"], True)
+        self.assert_is_sqlite("basic2")
+        assert_equal(basic2.getbalance(), basic2_balance)
+        self.assert_list_txs_equal(basic2.listtransactions(), basic2_txs)
+
+    def test_multisig(self):
+        default = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
+
+        # Contrived case where all the multisig keys are in a single wallet
+        self.log.info("Test migration of a wallet with all keys for a multisig")
+        multisig0 = self.create_legacy_wallet("multisig0")
+        addr1 = multisig0.getnewaddress()
+        addr2 = multisig0.getnewaddress()
+        addr3 = multisig0.getnewaddress()
+
+        ms_info = multisig0.addmultisigaddress(2, [addr1, addr2, addr3])
+
+        multisig0.migratewallet()
+        assert_equal(multisig0.getwalletinfo()["descriptors"], True)
+        self.assert_is_sqlite("multisig0")
+        ms_addr_info = multisig0.getaddressinfo(ms_info["address"])
+        assert_equal(ms_addr_info["ismine"], True)
+        assert_equal(ms_addr_info["desc"], ms_info["descriptor"])
+        assert_equal("multisig0_watchonly" in self.nodes[0].listwallets(), False)
+        assert_equal("multisig0_solvables" in self.nodes[0].listwallets(), False)
+
+        pub1 = multisig0.getaddressinfo(addr1)["pubkey"]
+        pub2 = multisig0.getaddressinfo(addr2)["pubkey"]
+
+        # Some keys in multisig do not belong to this wallet
+        self.log.info("Test migration of a wallet that has some keys in a multisig")
+        self.nodes[0].createwallet(wallet_name="multisig1")
+        multisig1 = self.nodes[0].get_wallet_rpc("multisig1")
+        ms_info = multisig1.addmultisigaddress(2, [multisig1.getnewaddress(), pub1, pub2])
+        ms_info2 = multisig1.addmultisigaddress(2, [multisig1.getnewaddress(), pub1, pub2])
+        assert_equal(multisig1.getwalletinfo()["descriptors"], False)
+
+        addr = ms_info["address"]
+        txid = default.sendtoaddress(addr, 10)
+        multisig1.importaddress(addr)
```

Comment:
fb5e0957e066bd0825947a399242242f1f1a2ae6: why do you need to call `importaddress` after `addmultisigaddress`? Is that makes it marked as `watch-only` (as oppose to the second address)? Why?

Reply:
So that this particular address is being watched and will end up in the `_watchonly` wallet.